agb/src/sound/mixer/mixer.s

```diff
@@ -0,0 +1,86 @@
+.arm
+.global agb_rs__mixer_add
+.section .iwram, "ax"
+.align
+agb_rs__mixer_add:
+    @ Arguments
+    @ r0 - pointer to the data to be copied (u8 array)
+    @ r1 - pointer to the sound buffer (i16 array)
+    @ r2 - playback speed (usize fixnum with 8 bits)
+    @ r3 - amount to modify the left channel by (u16 fixnum with 4 bits)
+    @ stack position 1 - amount to modify the right channel by (u16 fixnum with 4 bits)
+    @
+    @ The sound buffer must be SOUND_BUFFER_SIZE * 2 in size = 176 * 2
+
+    @ r9 = amount to modify right channel by
+
+    push {r4-r10}
+
+    ldr r9, [sp, #28]        @ load the right channel modification amount into r9
+
+    mov r12, #0              @ current write offset into the resulting buffer
+    mov r8, #352             @ the offset for writing to the resulting buffer between left and right channels
+
+    mov r5, #0               @ current index we're reading from
+
+    @ kept between iterations:
+    @ r12 - current write offset into the output buffer (r1)
+    @ r9  - the amount to modify the right channel by
+    @ r8  - the constant 352
+    @ r5  - the current index from the input buffer we're reading from (r0)
+    @ the provided arguments are all unmodified
+
+    @ all other registers are temporary
+1:
+    add r4, r0, r5, asr #8   @ calculate the address of the next read form the sound buffer
+    ldrsb r10, [r4]          @ load the current value we want to read
+    add r5, r5, r2           @ calculate the position to read the next step from
+
+    mov r6, r1               @ r6 = current buffer location
+
+    ldrh r4, [r6, r12]!      @ load the current buffer value (r12 being the offset) but pre-increment r6 by r12
+    mla r7, r10, r3, r4      @ r7 = r10 * r3 + r9 = current sound value * left amount + previous buffer value
+    strh r7, [r6], r8        @ *r6 = r7, r6 += r8 where r8 = 352 = offset for the right hand side
```

Comment:
inconsistent style compared to line 41?

Reply:
what do you mean?

Reply:
oh, you mean it should be `strh r7, [r6, r8]`?

Reply:
Yeah.

Reply:
You seem to use both styles, which just seems a bit odd to me, but who am I to judge.